go-controller/pkg/ovn/egressip.go

```diff
@@ -745,6 +766,84 @@ func findReroutePolicyIDs(filterOption, egressIPName string, gatewayRouterIP net
 	return strings.Split(policyIDs, "\n"), nil
 }
 
+func (oc *Controller) checkEgressNodesReachability() {
+	for {
+		reAddOrDelete := map[string]bool{}
+		oc.eIPC.allocatorMutex.Lock()
+		for _, eNode := range oc.eIPC.allocator {
+			if eNode.isEgressAssignable && eNode.isReady {
+				wasReachable := eNode.isReachable
+				isReachable := oc.isReachable(eNode)
+				if wasReachable && !isReachable {
+					reAddOrDelete[eNode.name] = true
+				} else if !wasReachable && isReachable {
+					reAddOrDelete[eNode.name] = false
+				}
+				eNode.isReachable = isReachable
+			}
+		}
+		oc.eIPC.allocatorMutex.Unlock()
+		for nodeName, shouldDelete := range reAddOrDelete {
+			node, err := oc.kube.GetNode(nodeName)
+			if err != nil {
+				klog.Errorf("Node: %s reachability changed, but could not retrieve node from API server, err: %v", node.Name, err)
+			}
+			if shouldDelete {
+				klog.Warningf("Node: %s is detected as unreachable, deleting it from egress assignment", node.Name)
+				if err := oc.deleteEgressNode(node); err != nil {
+					klog.Errorf("Node: %s is detected as unreachable, but could not re-assign egress IPs, err: %v", node.Name, err)
+				}
+			} else {
+				klog.Infof("Node: %s is detected as reachable and ready again, adding it to egress assignment", node.Name)
+				if err := oc.addEgressNode(node); err != nil {
```

Comment:
So again, the different conditions are not being merged correctly; in this case if the node has gone non-Ready (or been unlabelled), we will now forcibly re-enable it anyway when it becomes reachable again.

I think maybe `addEgressNode`/`deleteEgressNode` should just be replaced by a single `updateEgressNode` that computes the current availability (`isReady && isReachable && isAvailable`) and compares it to the previous state and either "adds" or "deletes" it (or does nothing) as needed.

Reply:
That's not quite right I believe. In this function we begin by checking if the nodes are `isAssignable` and `isReady`: https://github.com/ovn-org/ovn-kubernetes/blob/563c1eab21c16733d11732a5e347e0407c462830/go-controller/pkg/ovn/egressip.go#L774 and then proceed to add/delete accordingly. That should take care of the case you've mentioned.